deprecate `use(Lazy)Query` `on(Error|Completed)` callbacks (#12340)
apollographql/apollo-client: src/react/types/deprecation.md

Reviewer: @Meschreiber would you perhaps be willing to take a look at this as well? We are deprecating some callbacks and want to provide some justification for doing so. Your feedback is always appreciated, especially since you're much better at English than I am, so your suggestions would probably make more sense than mine 😆

While Apollo Client shares similar concerns, there are some additional reasons behind this change.

## `onCompleted`

### Conflicting interpretations for `onCompleted` behavior

Apollo Client uses a normalized cache, which means that there are many different
reasons why the data displayed by a component might change:

* The query initiated by the current hook might return data
* The same query might be initiated by another hook and update existing data
* You call `fetchMore`, `refetch`, etc.
* Another query or mutation might be overlapping with the query of the current hook and
  update some of its data
* An optimistic update might update some or all of the data of the hook
* A manual cache update might change the data for the hook

Apollo Client users have provided logical justification for each of these cases for why the `onCompleted` callback should or should not execute.

> ### Added ambiguisity around `@defer`

Author: Oh my god great catch, thank you @hwillson 🤣 
```suggestion
### Added ambiguity around `@defer`
```

What makes matters worse is that with the introduction of `@defer` in the GraphQL
ecosystem, we have yet another source of "updates".

> Should `onComplete` run once the initial chunk of data arrives? After all

Author: ```suggestion
Should `onCompleted` run once the initial chunk of data arrives? After all
```

deferred fragments arrived? After each fragment?
While one behavior might make sense to some, others might have vastly different conflicting opinions that are equally valid.

### Changes around the behaviour

> Adding insult to injury, we actually changed the behaviour of `onCompleted` slightly
> at one point - where previously the callback was always called whenever a result
> came in, now it's only called if a *different* result came in.
> This can be alleviated by setting `notifyOnNetworkStatusChange` to `true`, but
> it unfortunately adds to the confusion surrounding the callback.

Author: ```suggestion
Adding insult to injury, `onCompleted` had a bug in versions from 3.5 to 3.7 
where not only changes from the network would cause `onCompleted` to run,
but also cache changes.
We fixed this back in 3.8 with #10229, but the damage was done: some users
now expect one behaviour while others in the meantime had gotten used to the
other behaviour.
```

Author: What about this?

Reviewer: I like that 👍

# Deprecation of `useQuery` and `useLazyQuery` lifecycle hooks

With the release of Apollo Client 3.13, we will be deprecating the `useQuery`

> and `useLazyQuery` lifecyclye hooks `onCompleted` and `onError`.

Author: Yeah, good call.

These lifecycle hooks have for a long time been the cause of confusion, bugs and
frustration for many developers, and with this step we are following other tools
like React Query in their removal.

React Query's maintainer Dominik Dorfmeister has already done a better job in
[this comprehensive blog post](https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose)
explaining their choice, so I'd highly recommend giving it a read.

That said, we are a different library and so we have some different reasons behind

> this. I'll try my best to sum them up:

Author: Yeah, I used the "I" because it will be an issue that I'll be posting, not really an "official statement", but if we can keep it out I'm happy :)

reasons why the data displayed by a component might change:

* The query initiated by the current hook might return data
* The same query might be initiated by another hook and update existing data
* You call `fetchMore`
* Another query or mutation might be overlapping with the query of the current hook and
  update some of it's data
* An optimistic update might update some or all of the data of the hook
* Any other manual cache modification might update the data of the hook

For each of these events, you will find developers that see it as intuitive that
the `onCompleted` callback should execute - or that it definitely should not execute.

### Added ambiguisity around `@defer`

What makes matters worse is that with the introduction of `@defer` in the GraphQL
ecosystem, we have yet another source of "updates".
Do you want `onComplete` to run once the initial chunk of data arrives? After all
deferred fragments arrived? After each fragment?
The answer to this might be obvious to you, but turn around and talk to another
developer, and their opinion might be wildly different from yours - and equally
valid.

### Changes around the behaviour

> Adding insult to injury, we actually changed the behaviour of `onCompleted` slightly

Author: I'm honestly not sure if I could identify all the explicit cases where we had changes.
There is https://github.com/apollographql/apollo-client/pull/10229, but I believe we had more changes than just that, right?

> Adding insult to injury, we actually changed the behaviour of `onCompleted` slightly

Reviewer: To my knowledge, that is the only one. That is the spicy one that has caused the controversy, so I think linking that should be enough.

Given this history, even if we came up with a "perfect" and "intuitive" new behaviour,
any further change to these callbacks would just cause more and more confusion
among our userbase.

### Bugs

The final straw that made us come to this decision was [this bug report](https://github.com/apollographql/apollo-client/issues/12316):
With the current implementation, in some circumstances, the `onCompleted` and
`onError` callbacks can be stale by one render - and there's no good way we can
prevent that from happening in a matter that won't introduce new bugs in case you're
using suspense in your App.
React's [`useEffectEvent`](https://react.dev/learn/separating-events-from-effects#declaring-an-effect-event)
hook would solve this problem for us, but that hook is still experimental and
even when it is available, it won't be backported to the old React versions
we are still supporting.
With the currentlt available primitives, fixing this might be possible in a very
hacky way, but given everything else, we want to move everybody off these callbacks
instead of pushing additional bundle size on all our users for a feature we
don't recommend to use in the first place.

> ### What to use instead

Author: Hmm, you clearly have a better idea what to write there than I do - could you maybe contribute that paragraph? 😅
